File: test/codegen/js_passthrough.dart
library js_passthrough;

/// @internal.PassThrough
import 'dart:js';

String typeof(JsObject o) native;

class Foo {}

doStuff(JsObject obj) {
  obj['prop'] = 'value';
  print(obj['prop']);
}

main(args) {
  var someName = 'Foo';
  print(context[someName]);

  var o = new JsObject(context['ConstructorWithoutArgs']);
  o = new JsObject(context['ConstructorWithArgs'], ["1", 2]);
  o = new JsObject(context['ConstructorWithArgList'], args);
  print(o.callMethod('methodWithoutArgs'));
  print(o.callMethod('methodWithArgs', [1, "2"]));
  print(o.callMethod('methodWithArgList', args));
  o.deleteProperty('foo');
  print(o.hasProperty('foo'));
  print(o.instanceof(context['Bar']));
  print(typeof(o));
  doStuff(new Foo() as JsObject);

  JsFunction f = context['Symbol'] as JsFunction;
  print(f.apply(['someSymbol']));
  print(f.apply(args));
  print(f.apply(args, thisArg: o));

  print(context['Symbol'].apply(['_runtimeType']));
}

File: test/codegen/expect/js_passthrough.js
dart_library.library('js_passthrough', null, /* Imports */[
  "dart_runtime/dart",
  'dart/core',
  'dart/js'
], /* Lazy imports */[
], function(exports, dart, core, js) {
  'use strict';
  let dartx = dart.dartx;
  class Foo extends core.Object {}
  function doStuff(obj) {
    obj.prop = 'value';
    core.print(obj.prop);
  }
  dart.fn(doStuff, dart.dynamic, [js.JsObject]);
  function main(args) {
    let someName = 'Foo';
    core.print(dart.global[someName]);
    let o = new ConstructorWithoutArgs();
    o = new ConstructorWithArgs("1", 2);
    o = new (Function.prototype.bind.apply(ConstructorWithArgList, [null, dart.as(args, core.List)]))();
    core.print(o.methodWithoutArgs());
    core.print(o.methodWithArgs(1, "2"));
    core.print(((target, args) => target.methodWithArgList.apply(target, args))(o, dart.as(args, core.List)));
    delete o.foo;
    core.print('foo' in o);
    core.print(o instanceof Bar);
    core.print(typeof(o));
    doStuff(new Foo());
    let f = Symbol;
    core.print(f('someSymbol'));
    core.print(f.apply(null, dart.as(args, core.List)));
    core.print(f.apply(o, dart.as(args, core.List)));
    core.print(Symbol('_runtimeType'));
  }
  dart.fn(main);
  // Exports:
  exports.Foo = Foo;
  exports.doStuff = doStuff;
  exports.main = main;
});
